File: Backend/routes/addresses.py
from fastapi import APIRouter, HTTPException, Depends, status
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional, List
from database import get_db
from middleware.auth import verify_token
from models import User, Address

router = APIRouter()

class CreateAddressRequest(BaseModel):
    label: str
    firstName: str
    lastName: str
    phone: str
    address: str
    city: str
    postalCode: str
    country: str = "France"
    isDefault: bool = False

class UpdateAddressRequest(BaseModel):
    label: Optional[str] = None
    firstName: Optional[str] = None
    lastName: Optional[str] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    postalCode: Optional[str] = None
    country: Optional[str] = None
    isDefault: Optional[bool] = None

@router.get("/")
async def get_addresses(
    user: User = Depends(verify_token),
    db: Session = Depends(get_db)
):
    addresses = db.query(Address).filter(Address.user_id == user.id).all()
    return [
        {
            "id": addr.id,
            "label": addr.label,
            "firstName": addr.first_name,
            "lastName": addr.last_name,
            "phone": addr.phone,
            "address": addr.address,
            "city": addr.city,
            "postalCode": addr.postal_code,
            "country": addr.country,
            "isDefault": addr.is_default,
            "createdAt": addr.created_at.isoformat() if addr.created_at else None
        }
        for addr in addresses
    ]

@router.post("/")
async def create_address(
    request: CreateAddressRequest,
    user: User = Depends(verify_token),
    db: Session = Depends(get_db)
):
    if request.isDefault:
        db.query(Address).filter(Address.user_id == user.id).update({"is_default": False})
    
    address = Address(
        user_id=user.id,
        label=request.label,
        first_name=request.firstName,
        last_name=request.lastName,
        phone=request.phone,
        address=request.address,
        city=request.city,
        postal_code=request.postalCode,
        country=request.country,
        is_default=request.isDefault
    )
    
    db.add(address)
    db.commit()
    db.refresh(address)
    
    return {
        "id": address.id,
        "label": address.label,
        "firstName": address.first_name,
        "lastName": address.last_name,
        "phone": address.phone,
        "address": address.address,
        "city": address.city,
        "postalCode": address.postal_code,
        "country": address.country,
        "isDefault": address.is_default
    }

@router.put("/{address_id}")
async def update_address(
    address_id: int,
    request: UpdateAddressRequest,
    user: User = Depends(verify_token),
    db: Session = Depends(get_db)
):
    address = db.query(Address).filter(Address.id == address_id, Address.user_id == user.id).first()
    if not address:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Adresse non trouvée")
    
    if request.isDefault is True:
        db.query(Address).filter(Address.user_id == user.id, Address.id != address_id).update({"is_default": False})
    
    if request.label is not None:
        address.label = request.label
    if request.firstName is not None:
        address.first_name = request.firstName
    if request.lastName is not None:
        address.last_name = request.lastName
    if request.phone is not None:
        address.phone = request.phone
    if request.address is not None:
        address.address = request.address
    if request.city is not None:
        address.city = request.city
    if request.postalCode is not None:
        address.postal_code = request.postalCode
    if request.country is not None:
        address.country = request.country
    if request.isDefault is not None:
        address.is_default = request.isDefault
    
    db.commit()
    db.refresh(address)
    
    return {
        "id": address.id,
        "label": address.label,
        "firstName": address.first_name,
        "lastName": address.last_name,
        "phone": address.phone,
        "address": address.address,
        "city": address.city,
        "postalCode": address.postal_code,
        "country": address.country,
        "isDefault": address.is_default
    }

@router.delete("/{address_id}")
async def delete_address(
    address_id: int,
    user: User = Depends(verify_token),
    db: Session = Depends(get_db)
):
    address = db.query(Address).filter(Address.id == address_id, Address.user_id == user.id).first()
    if not address:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Adresse non trouvée")
    
    db.delete(address)
    db.commit()
    return {"message": "Adresse supprimée"}

File: Backend-nodejs/routes/addresses.js
import express from 'express';
import { query } from '../config/database.js';
import { verifyToken } from '../middleware/auth.js';
import { validateAddressId, validateCreateAddress, handleValidationErrors } from '../middleware/validation.js';
import { publicRateLimiter } from '../config/security.js';

const router = express.Router();

// Toutes les routes nécessitent une authentification
router.use(verifyToken);
router.use(publicRateLimiter);

/**
 * GET /api/addresses
 * Récupère toutes les adresses de l'utilisateur
 */
router.get('/', async (req, res) => {
  try {
    const userId = req.user.id;

    const addresses = await query(
      `SELECT id, label, first_name, last_name, phone, address, city, postal_code, country, is_default, created_at
       FROM addresses 
       WHERE user_id = ? 
       ORDER BY is_default DESC, created_at DESC`,
      [userId]
    );

    res.json(addresses.map(addr => ({
      id: addr.id,
      label: addr.label,
      firstName: addr.first_name,
      lastName: addr.last_name,
      phone: addr.phone,
      address: addr.address,
      city: addr.city,
      postalCode: addr.postal_code,
      country: addr.country,
      isDefault: addr.is_default === 1 || addr.is_default === true,
      createdAt: addr.created_at ? new Date(addr.created_at).toISOString() : null
    })));
  } catch (error) {
    console.error('Erreur lors de la récupération des adresses:', error);
    res.status(500).json({
      error: 'Erreur serveur',
      detail: 'Erreur lors de la récupération des adresses'
    });
  }
});

/**
 * POST /api/addresses
 * Crée une nouvelle adresse
 */
router.post('/', validateCreateAddress, async (req, res) => {
  try {
    const userId = req.user.id;
    const { label, firstName, lastName, phone, address, city, postalCode, country = 'France', isDefault = false } = req.body;

    // Si c'est l'adresse par défaut, désactiver les autres
    if (isDefault) {
      await query(
        'UPDATE addresses SET is_default = 0 WHERE user_id = ?',
        [userId]
      );
    }

    const result = await query(
      `INSERT INTO addresses (user_id, label, first_name, last_name, phone, address, city, postal_code, country, is_default, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW(), NOW())`,
      [userId, label, firstName, lastName, phone, address, city, postalCode, country, isDefault ? 1 : 0]
    );

    const newAddress = await query(
      'SELECT id, label, first_name, last_name, phone, address, city, postal_code, country, is_default FROM addresses WHERE id = ?',
      [result.insertId]
    );

    res.status(201).json({
      id: newAddress[0].id,
      label: newAddress[0].label,
      firstName: newAddress[0].first_name,
      lastName: newAddress[0].last_name,
      phone: newAddress[0].phone,
      address: newAddress[0].address,
      city: newAddress[0].city,
      postalCode: newAddress[0].postal_code,
      country: newAddress[0].country,
      isDefault: newAddress[0].is_default === 1 || newAddress[0].is_default === true
    });
  } catch (error) {
    console.error('Erreur lors de la création de l\'adresse:', error);
    res.status(500).json({
      error: 'Erreur serveur',
      detail: 'Erreur lors de la création de l\'adresse'
    });
  }
});

/**
 * PUT /api/addresses/:address_id
 * Met à jour une adresse
 */
router.put('/:address_id', validateAddressId, async (req, res) => {
  try {
    const userId = req.user.id;
    const addressId = parseInt(req.params.address_id);
    const { label, firstName, lastName, phone, address, city, postalCode, country, isDefault } = req.body;

    // Vérifier que l'adresse appartient à l'utilisateur
    const existingAddresses = await query(
      'SELECT id FROM addresses WHERE id = ? AND user_id = ?',
      [addressId, userId]
    );

    if (!existingAddresses || existingAddresses.length === 0) {
      return res.status(404).json({
        error: 'Erreur',
        detail: 'Adresse non trouvée'
      });
    }

    // Si c'est l'adresse par défaut, désactiver les autres
    if (isDefault === true) {
      await query(
        'UPDATE addresses SET is_default = 0 WHERE user_id = ? AND id != ?',
        [userId, addressId]
      );
    }

    const updates = [];
    const values = [];

    if (label !== undefined) {
      updates.push('label = ?');
      values.push(label);
    }
    if (firstName !== undefined) {
      updates.push('first_name = ?');
      values.push(firstName);
    }
    if (lastName !== undefined) {
      updates.push('last_name = ?');
      values.push(lastName);
    }
    if (phone !== undefined) {
      updates.push('phone = ?');
      values.push(phone);
    }
    if (address !== undefined) {
      updates.push('address = ?');
      values.push(address);
    }
    if (city !== undefined) {
      updates.push('city = ?');
      values.push(city);
    }
    if (postalCode !== undefined) {
      updates.push('postal_code = ?');
      values.push(postalCode);
    }
    if (country !== undefined) {
      updates.push('country = ?');
      values.push(country);
    }
    if (isDefault !== undefined) {
      updates.push('is_default = ?');
      values.push(isDefault ? 1 : 0);
    }

    if (updates.length === 0) {
      return res.status(400).json({
        error: 'Erreur',
        detail: 'Aucune donnée à mettre à jour'
      });
    }

    updates.push('updated_at = NOW()');
    values.push(addressId);

    await query(
      `UPDATE addresses SET ${updates.join(', ')} WHERE id = ?`,
      values
    );

    const updatedAddress = await query(
      'SELECT id, label, first_name, last_name, phone, address, city, postal_code, country, is_default FROM addresses WHERE id = ?',
      [addressId]
    );

    res.json({
      id: updatedAddress[0].id,
      label: updatedAddress[0].label,
      firstName: updatedAddress[0].first_name,
      lastName: updatedAddress[0].last_name,
      phone: updatedAddress[0].phone,
      address: updatedAddress[0].address,
      city: updatedAddress[0].city,
      postalCode: updatedAddress[0].postal_code,
      country: updatedAddress[0].country,
      isDefault: updatedAddress[0].is_default === 1 || updatedAddress[0].is_default === true
    });
  } catch (error) {
    console.error('Erreur lors de la mise à jour de l\'adresse:', error);
    res.status(500).json({
      error: 'Erreur serveur',
      detail: 'Erreur lors de la mise à jour de l\'adresse'
    });
  }
});

/**
 * DELETE /api/addresses/:address_id
 * Supprime une adresse
 */
router.delete('/:address_id', validateAddressId, async (req, res) => {
  try {
    const userId = req.user.id;
    const addressId = parseInt(req.params.address_id);

    // Vérifier que l'adresse appartient à l'utilisateur
    const existingAddresses = await query(
      'SELECT id FROM addresses WHERE id = ? AND user_id = ?',
      [addressId, userId]
    );

    if (!existingAddresses || existingAddresses.length === 0) {
      return res.status(404).json({
        error: 'Erreur',
        detail: 'Adresse non trouvée'
      });
    }

    await query('DELETE FROM addresses WHERE id = ?', [addressId]);

    res.json({
      message: 'Adresse supprimée'
    });
  } catch (error) {
    console.error('Erreur lors de la suppression de l\'adresse:', error);
    res.status(500).json({
      error: 'Erreur serveur',
      detail: 'Erreur lors de la suppression de l\'adresse'
    });
  }
});

export default router;
